internals/overlord/checkstate/manager.go

```diff
@@ -48,42 +87,114 @@ func (m *CheckManager) NotifyCheckFailed(f FailureFunc) {
+func (m *CheckManager) PlanChanged(newPlan *plan.Plan) {
+	m.state.Lock()
+	defer m.state.Unlock()
+
+	// Update local reference to plan.
+	m.planLock.Lock() // always acquire locks in same order (state lock, then plan lock)
+	oldPlan := m.plan
+	m.plan = newPlan
+	m.planLock.Unlock()
+
+	if oldPlan == nil {
+		oldPlan = &plan.Plan{}
+	}
+	shouldEnsure := false
+	newOrModified := make(map[string]bool, len(newPlan.Checks))
+
+	// Abort all currently-running checks.
+	for _, change := range m.state.Changes() {
+		switch change.Kind() {
+		case performCheckKind, recoverCheckKind:
+			if change.IsReady() {
+				// Skip check changes that have finished already.
+				continue
+			}
+			details := mustGetCheckDetails(change)
+			oldConfig, inOld := oldPlan.Checks[details.Name]
+			newConfig, inNew := newPlan.Checks[details.Name]
+			if inOld && inNew {
+				if reflect.DeepEqual(oldConfig, newConfig) {
+					// Don't restart check if its configuration hasn't changed.
+					continue
+				}
+				// Check is in old and new plans and has been modified.
+				newOrModified[details.Name] = true
+			}
+			change.Abort()
+			m.deleteHealthInfo(details.Name)
+			shouldEnsure = true
+		}
+	}
 
-	logger.Debugf("Configuring check manager (stopping %d, starting %d)",
-		len(m.checks), len(p.Checks))
+	// Also find checks that are new (in new plan but not in old one).
+	for _, config := range newPlan.Checks {
+		if oldPlan.Checks[config.Name] == nil {
+			newOrModified[config.Name] = true
+		}
+	}
 
-	// First stop existing checks.
-	for _, check := range m.checks {
-		check.cancel()
+	// Start new or modified checks.
+	for _, config := range newPlan.Checks {
+		if newOrModified[config.Name] {
+			performCheckChange(m.state, config)
+			m.updateHealthInfo(config, 0)
+			shouldEnsure = true
+		}
+	}
+	if !m.ensureDone.Load() {
+		// Can't call EnsureBefore before Overlord.Loop is running (which will
+		// call m.Ensure for the first time).
+		return
+	}
+	if shouldEnsure {
+		m.state.EnsureBefore(0) // start new tasks right away
 	}
-	// Wait for all context cancellations to propagate and allow
-	// each goroutine to cleanly exit.
-	m.wg.Wait()
-
-	// Set the size of the next wait group
-	m.wg.Add(len(p.Checks))
-
-	// Then configure and start new checks.
-	checks := make(map[string]*checkData, len(p.Checks))
-	for name, config := range p.Checks {
-		ctx, cancel := context.WithCancel(context.Background())
-		check := &checkData{
-			config:  config,
-			checker: newChecker(config, p),
-			ctx:     ctx,
-			cancel:  cancel,
-			action:  m.callFailureHandlers,
+}
+
+func (m *CheckManager) changeStatusChanged(chg *state.Change, old, new state.Status) {
+	// Always acquire locks in same order (state lock, then plan lock).
+	// The state engine has already acquired the state lock at this point.
+	m.planLock.Lock()
+	plan := m.plan
+	m.planLock.Unlock()
+
+	shouldEnsure := false
+	switch {
+	case chg.Kind() == performCheckKind && new == state.ErrorStatus:
```

Comment:
I cannot remember what the final state of an Aborted change is? I am wondering what will happen with a check that is detected as modified, where the original change will be aborted. If that happens to be an ```performCheckKind``` change, will it end up here?

Reply:
Aborted changes get moved to Done status, so this one be okay, however, the `chg.Kind() == recoverCheckKind && new == state.DoneStatus` case below will yield true. But that's why I have the `details.Proceed` check in addition: I set a "proceed=true" boolean in the task data when the change should proceed to the next change kind. That's also useful if the task returns for some other reason (eg: perform-check returning another kind of error).